language/resource-api/README.md

```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
       is = change.has_key? :is ? change[:is] : get_single(name)
       should = change[:should]
       # ...
-      context.device.do_something unless noop
+      if context.transport.verify
+        context.transport.do_something unless noop
+      else
+        raise Puppet::Transport::Error('Connection to remote target is not available')
+      end
     end
   end
 end
 ```
 
-Declaring this feature restricts the resource from being run "locally". It is expected to execute all external interactions through the `context.device` instance. The way that instance is set up is runtime specific. In Puppet, it is configured through the [`device.conf`](https://puppet.com/docs/puppet/5.3/config_file_device.html) file, and only available when running under [`puppet device`](https://puppet.com/docs/puppet/5.3/man/device.html). It is recommended to use `Puppet::Util::NetworkDevice::Simple::Device` as the base class for all devices, which automatically loads a configuration from the local filesystem of the proxy node where it is running on.
+Declaring this feature restricts the resource from being run "locally". It is expected to execute all external interactions through the `context.transport` instance. The way that instance is set up is runtime specific. In Puppet, it is configured through the [`device.conf`](https://puppet.com/docs/puppet/5.3/config_file_device.html) file, and only available when running under [`puppet device`](https://puppet.com/docs/puppet/5.3/man/device.html).
+
+### Transport
+
+```ruby
+# lib/puppet/transport/nexus_schema.rb
+Puppet::ResourceAPI.register_transport(
+  name: 'nexus', # points at class Puppet::Transport::Nexus
+  desc: 'Connects to a Cisco Nexus device'
+  features: [], # TODO
+  connection_info: {
+    connector_type: {
+      type: 'Enum[nxapi, ssh, grpc, pidgeon]',
+      desc: 'Which method to use to connect to the host'
+    },
+    hostname: {
+      type: 'String',
+      desc: 'The host to connect to.',
+    },
+    username: {
+      type: 'String',
+      desc: 'The user.',
+    },
+    password: {
+      type: 'Sensitive[String]',
+      desc: 'The password to connect.',
+    },
+    enable_password: {
+      type: 'Sensitive[String]',
+      desc: 'The password escalate to enable access.',
+    },
+    port: {
+      type: 'Integer',
+      desc: 'The port to connect to.',
+    },
+  },
+)
+
+#  lib/puppet/transport/nexus.rb
+class Puppet::Transport::Nexus
+  def initialize(transport_info); end
+  def facts; end
+  def verify; end
+  def close; end
+end
+```
+
+The transport layer consists of the schema and the implementation. The schema is defined in the same manner as a `Type`, except instead of `attributes`,
+ there is now `connection_info` and it describes the shape of the data which should be passed to the implementation for a connection to be made.
+
+The implementation must follow the structure as defined above and provide the following methods:
+
+  * initialize(connection_info)
```

Comment:
Is initialize expected to open the connection?

Reply:
Added more extensive explanation to the doc